exporter/loadbalancingexporter/resolver_k8s.go

```diff
@@ -0,0 +1,301 @@
+
+	close(r.stopCh)
+	r.shutdownWg.Wait()
+	return nil
+}
+func newInClusterClient() (kubernetes.Interface, error) {
+	cfg, err := config.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	return kubernetes.NewForConfig(cfg)
+}
+
+func (r *k8sResolver) resolve(ctx context.Context) ([]string, error) {
+	r.shutdownWg.Add(1)
+	defer r.shutdownWg.Done()
+
+	var backends []string
+	r.endpointsStore.Range(func(address, value any) bool {
+		addr := address.(string)
+		var backend string
+		ip, err := net.ResolveIPAddr("ip", addr)
+		if err != nil {
+			stats.RecordWithTags(ctx, k8sResolverSuccessFalseMutators, mNumResolutions.M(1))
+			r.logger.Error("resolve address error", zap.String("address", addr), zap.Error(err))
+			return true
+		}
+		stats.RecordWithTags(ctx, k8sResolverSuccessTrueMutators, mNumResolutions.M(1))
+		if ip.IP.To4() != nil {
+			backend = ip.String()
+		} else {
+			// it's an IPv6 address
+			backend = fmt.Sprintf("[%s]", ip.String())
+		}
+
+		for _, port := range r.port {
+			// if a port is specified in the configuration, add it
+			if port != 0 {
+				backends = append(backends, fmt.Sprintf("%s:%d", backend, port))
+			}
+		}
+		return true
+	})
+
+	// keep it always in the same order
+	sort.Strings(backends)
+
+	if equalStringSlice(r.endpoints, backends) {
+		return r.endpoints, nil
+	}
+
+	// the list has changed!
+	r.updateLock.Lock()
+	r.endpoints = backends
+	r.updateLock.Unlock()
+	_ = stats.RecordWithTags(ctx, resolverSuccessTrueMutators, mNumBackends.M(int64(len(backends))))
+
+	// propagate the change
+	r.changeCallbackLock.RLock()
+	for _, callback := range r.onChangeCallbacks {
+		callback(r.endpoints)
+	}
+	r.changeCallbackLock.RUnlock()
+	return r.endpoints, nil
+}
+
+func (r *k8sResolver) onChange(f func([]string)) {
+	r.changeCallbackLock.Lock()
+	defer r.changeCallbackLock.Unlock()
+	r.onChangeCallbacks = append(r.onChangeCallbacks, f)
+}
+
+var _ cache.ResourceEventHandler = (*handler)(nil)
+
+type handler struct {
+	endpoints *sync.Map
+	callback  func(ctx context.Context) ([]string, error)
+	logger    *zap.Logger
+}
+
+func (h handler) OnAdd(obj interface{}, isInInitialList bool) {
+	h.logger.Debug("onAdd called")
+	var endpoints []string
+	switch object := obj.(type) {
+	case *corev1.Endpoints:
+		endpoints = convertToEndpoints(object)
+	default: // unsupported
+		return
+	}
+	changed := false
+	for _, ep := range endpoints {
+		if _, loaded := h.endpoints.LoadOrStore(ep, ep); !loaded {
+			changed = true
+		}
+	}
+	h.logger.Debug("onAdd check", zap.Bool("changed", changed))
+	if changed {
+		h.callback(context.TODO())
+	}
+}
+
+func (h handler) OnUpdate(oldObj, newObj interface{}) {
+	h.logger.Debug("onUpdate called")
+	switch oldObj.(type) {
+	case *corev1.Endpoints:
+		newEps, ok := newObj.(*corev1.Endpoints)
+		if !ok {
+			return
+		}
+		endpoints := convertToEndpoints(newEps)
+		changed := false
+		for _, ep := range endpoints {
+			if _, loaded := h.endpoints.LoadOrStore(ep, ep); !loaded {
```

Comment:
Instead of storing a map[string]string, you could store a map[string]bool, and set this value here to true. You'd save some space with that, while still having the semantics you need.

Reply:
done